pkg/kwokctl/runtime/compose/cluster.go

```diff
@@ -46,15 +46,32 @@ import (
 // Cluster is a implementation of Runtime for docker-compose.
 type Cluster struct {
 	*runtime.Cluster
+
+	driver string
 }
 
-// NewCluster creates a new Runtime for docker-compose.
-func NewCluster(name, workdir string) (runtime.Runtime, error) {
+// NewNerdctlCluster creates a new Runtime for nerdctl compose.
+func NewNerdctlCluster(name, workdir string) (runtime.Runtime, error) {
 	return &Cluster{
 		Cluster: runtime.NewCluster(name, workdir),
+		driver:  consts.RuntimeTypeNerdctl,
 	}, nil
 }
 
+// NewDockerCluster creates a new Runtime for docker-compose.
+func NewDockerCluster(name, workdir string) (runtime.Runtime, error) {
+	return &Cluster{
+		Cluster: runtime.NewCluster(name, workdir),
+		driver:  consts.RuntimeTypeDocker,
+	}, nil
+}
+
+// Available  checks whether the runtime is available.
+func (c *Cluster) Available(ctx context.Context) error {
+	return exec.Exec(context.Background(), "", exec.IOStreams{},
+		c.driver, "version")
+}
+
```

Comment:
```suggestion
// NewCluster creates a new Runtime for docker-compose.
func NewCluster(name, workdir string) (runtime.Runtime, error) {
	return &Cluster{
		Cluster: runtime.NewCluster(name, workdir),
	}, nil
}

// Available  checks whether the runtime is available.
func (c *Cluster) Available(ctx context.Context) error {
	conf, err := c.Config(ctx)
	if err != nil {
		return err
	}

	return exec.Exec(ctx, "", exec.IOStreams{}, config.Options.Runtime, "version")
}

```

Revert other changes in the file

Reply:
`cluster.Available` is used to detect whether the runtime is available, at this time, `config.Options.Runtime` is empty, it does not know it is `docker` or `nerdctl`

Reply:
we exposed `docker` and `nerdctl` as 2 runtimes, IMHO, It is reasonable to offer `NewDockerCluster` and `NewNerdctlCluster`, those 2 runtimes can know what they are and reuse this `compose` logic in most places.

Reply:
> `cluster.Available` is used to detect whether the runtime is available, at this time, `config.Options.Runtime` is empty, it does not know it is `docker` or `nerdctl`

After creating the cluster, the Runtime is already set and cannot be empty here.

> we exposed `docker` and `nerdctl` as 2 runtimes, IMHO, It is reasonable to offer `NewDockerCluster` and `NewNerdctlCluster`, those 2 runtimes can know what they are and reuse this `compose` logic in most places.

Maybe if we add a parameter runtime to NewCluster, this should be more appropriate and we can leave it unchanged in the future when we add podman-compose. WDYT?

Reply:
> After creating the cluster, the Runtime is already set and cannot be empty here.

check the logic you suggested https://github.com/kubernetes-sigs/kwok/pull/244#discussion_r1095454530, the runtime read from config is set AFTER `rt.Available()`, we will always read an empty runtime from config in `rt.Available()` when we trying to detect runtime automatically.

the reason I added a `driver` param to the compost Cluster is to set it after creating the cluster

> Maybe if we add a parameter runtime to NewCluster
NewCluster only accepts `name` and `workdir`, it is odd if we add a `runtime-kind` parameter to `NewCluster` when we already specified we are creating some kind of runtime like `compose.NewCluster(name, workdir, docker)`, `kind.NewCluster(name, workdir, kind)`

providing `NewNerdctlCluster` and `NewDockerCluster` and registering it to the registry seems more elegant.

> this should be more appropriate and we can leave it unchanged in the future when we add podman-compose.

this is not that possible we can add `podman` support without adding some incompatibility like currently existed `if runtime == Nerdctl`,

we only have to provide a function `NewPodmanCluster` when Kwok supports Podman if we follow the `NewNerdctlCluster`, `NewDockerCluster` way, and maintain the incompatibility inside compose package.

Reply:
That's fine, but this part of the refactoring is not necessary in this feature, so maybe we can discuss this in the new PR

Reply:
I can raise another PR to discuss `NewCluster`, but the there is a problem need to be solved, let me paste a screenshot to explain myself:

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/3764335/216778446-93366e42-924e-4341-a014-5fc587306578.png">

we can refactor it like this, but it seemed odd

<img width="720" alt="image" src="https://user-images.githubusercontent.com/3764335/216778520-b69d1d30-fa7e-458e-9f77-af066ec1ee0b.png">

<img width="1360" alt="image" src="https://user-images.githubusercontent.com/3764335/216778609-6c92e55b-5b63-4298-a4b1-67ce64a430f6.png">


that is mainly why I make the change

Reply:
Yeah, You're right, do as you say.